File: Language/C#/Book/tmp/LINQ实战/LinqInAction/Chapter12.VB/LinqToAmazon/AmazonHelper.vb
Imports System.Web
Imports System.Globalization
Imports AmazonCommon

NotInheritable Class AmazonHelper
  Private Sub New()
  End Sub

  Friend Shared Function BuildUrl(ByVal criteria As AmazonBookQueryCriteria) As String
    If criteria Is Nothing Then
      Throw New ArgumentNullException("criteria")
    End If

    Dim signHelper = New SignedRequestHelper(Amazon.AccessKey, Amazon.SecretKey, "ecs.amazonaws.com")
    Dim parameters = New Dictionary(Of String, String)()
    parameters("Service") = Amazon.ServiceName
    parameters("Version") = Amazon.ServiceVersion
    parameters("Operation") = "ItemSearch"
    parameters("SearchIndex") = "Books"
    parameters("ResponseGroup") = "Medium"

    If Not [String].IsNullOrEmpty(criteria.Title) Then
      parameters("Title") = criteria.Title
    End If
    If Not [String].IsNullOrEmpty(criteria.Publisher) Then
      parameters("Publisher") = criteria.Publisher
    End If
    If criteria.Condition.HasValue Then
      parameters("Condition") = criteria.Condition.ToString()
    End If
    If criteria.MaximumPrice.HasValue Then
      parameters("MaximumPrice") = (criteria.MaximumPrice * 100).Value.ToString(CultureInfo.InvariantCulture)
    End If

    Dim url = signHelper.Sign(parameters)
    Return url
  End Function

  Friend Shared Function PerformWebQuery(ByVal url As String) As IEnumerable(Of AmazonBook)
    ' Execute query
    Dim booksDoc As XElement = XElement.Load(url)

    ' Parse results
    Dim ns As XNamespace = Amazon.AmazonNS
    Dim books As IEnumerable(Of AmazonBook) = _
      From book In booksDoc.Descendants(ns + "Item") _
      Let attributes = book.Element(ns + "ItemAttributes") _
      Let price = attributes.Element(ns + "ListPrice").Element(ns + "Amount").Value _
      Let isbn = attributes.Element(ns + "ISBN") _
      Let asin = book.Element(ns + "ASIN") _
      Select New AmazonBook With _
           { _
             .Title = attributes.Element(ns + "Title").Value, _
             .Isbn = If(Not isbn Is Nothing, isbn.Value, asin.Value), _
             .PageCount = UInt32.Parse(attributes.Element(ns + "NumberOfPages").Value), _
             .Price = If(Not price Is Nothing, Decimal.Parse(price) / 100, 0), _
             .Publisher = attributes.Element(ns + "Publisher").Value, _
             .Year = UInt32.Parse(CType(attributes.Element(ns + "PublicationDate").Value, String).Substring(0, 4)), _
             .Authors = ( _
                From author In book.Descendants(ns + "Author") _
                Select CType(author.Value, String) _
              ).ToList() _
           }

    Return books
  End Function
End Class

File: Language/C#/Book/LINQ实战/第12章 扩展LINQ/Chapter12.CSharp/LinqToAmazon/AmazonHelper.cs
#region Imports

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Web;
using System.Xml.Linq;
using AmazonCommon;

#endregion Imports

namespace LinqInAction.Chapter12.LinqToAmazon
{
  static class AmazonHelper
  {
    static internal String BuildUrl(AmazonBookQueryCriteria criteria)
    {
      if (criteria == null)
        throw new ArgumentNullException("criteria");

      var signHelper = new SignedRequestHelper(Amazon.AccessKey, Amazon.SecretKey, "ecs.amazonaws.com");
      var parameters = new Dictionary<String, String>() {
        { "Service", Amazon.ServiceName},
        { "Version", Amazon.ServiceVersion },
        { "Operation", "ItemSearch" },
        { "SearchIndex", "Books" },
        { "ResponseGroup", "Medium" }
      };

      if (!String.IsNullOrEmpty(criteria.Title))
        parameters["Title"] = criteria.Title;
      if (!String.IsNullOrEmpty(criteria.Publisher))
        parameters["Publisher="] = criteria.Publisher;
      if (criteria.Condition.HasValue)
        parameters["Condition="] = criteria.Condition.ToString();
      if (criteria.MaximumPrice.HasValue)
        parameters["MaximumPrice"] = (criteria.MaximumPrice * 100).Value.ToString(CultureInfo.InvariantCulture);

      String url = signHelper.Sign(parameters);
      return url;
    }

    static internal IEnumerable<AmazonBook> PerformWebQuery(String url)
    {
      // Execute query
      XElement booksDoc = XElement.Load(url);

      // Parse results
      XNamespace ns = Amazon.AmazonNS;
      IEnumerable<AmazonBook> books =
        from book in booksDoc.Descendants(ns + "Item")
        let attributes = book.Element(ns + "ItemAttributes")
        let price = attributes.Element(ns + "ListPrice").Element(ns + "Amount").Value
        let isbn = attributes.Element(ns + "ISBN")
        let asin = book.Element(ns + "ASIN")
        select new AmazonBook
           {
             Title = attributes.Element(ns + "Title").Value,
             Isbn = (isbn ?? asin).Value,
             PageCount = UInt32.Parse(attributes.Element(ns + "NumberOfPages").Value),
             Price = price != null ? Decimal.Parse(price) / 100 : 0,
             Publisher = attributes.Element(ns + "Publisher").Value,
             Year = UInt32.Parse(
               ((String)attributes.Element(ns + "PublicationDate").Value).Substring(0, 4)),
             Authors = (
               from author in book.Descendants(ns + "Author")
               select (String)author.Value
             ).ToList()
           };

      return books;
    }
  }
}